### PR TITLE
LPS-121610 Creates a CSSVariablesProvider for default Theme token values

### PR DESCRIPTION
**Pull request by jbalsas**
As it's stated in the commit:

> To avoid having to hardcode default token values in the css implementation and simplify the generation of new tokens, this CSSVariablesProvider provides default values for all tokens in a token definition with a cssVariable mapping.
> 
> The variable definitions are pushed to the ":root" element so they apply to the whole page and can be overridden in context using css scopes.

**The Problem**

To generate a new token right now we need:
- Update the token defintion (in `token-definition.json`)
- Provide CSS implementation for the token

Right now, because we don't push default token value definitions anywhere, the developer is forced to add a manual fallback in CSS duplicating the token default value both in the `css` and in the `token-definition.json` as:

```css
.header-bg {
	background-color: $headerBackground;
	background-color: var(--header-background, $headerBackground);
}
```

**The Fix**

Provide a `CSSVariablesProvider` to inject the page's default token values in the `:root` scope.

**Test Plan**

- Build this thing!
- Deploy this thing!
- Assert that the css variables with default values are applied to the `:root` scope

Screen Shot 2020-10-01 at 14 32 30

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @izaera, I just hacked this together (just wanted to code something). The Java is awful and the names suck and I doubt it will format ever 😂 

Could you please fix it? 🙏

**Comment by jbalsas** [comment]
Hey @liferay-echo, just in case you want to take a look at this!

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Chema Balsas
+ */
+@Component(service = ScopedCSSVariablesProvider.class)
```
Comment (izaera) [inline modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:44]: I guess this is missing a `service.ranking` so that the precedence between it and the provider from `style-book-service` is well defined :thinking:

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:44]: Unless the CSS is already taking care of that?

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:44]: This is is pushed to `:root` whereas Stylebooks go to `#wrapper`. I think that should be enough for now, but it won't hurt giving this a priority service ranking if we want to be sure.

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:44]: I think I'll add the `service.ranking` just to make sure it doesn't break in the future if we change the selectors...

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:44]: Sounds good, thanks!

**Comment by izaera** [comment]
I'm going to turn this into a parser for Token Definitions JSON files because I can't find a better way to format the code and it's something that will be probably needed, in fact.

I will push the commits with the proposal and we can discuss it over here.

**Comment by victorg1991** [comment]
We'll benefit from the parser too, as we have a couple of places where we are doing the ugly iteration over the JSON structure in java :D

**Comment by jbalsas** [comment]
> I'm going to turn this into a parser for Token Definitions JSON files because I can't find a better way to format the code and it's something that will probably be needed, in fact.

goforit

**Comment by izaera** [comment]
I have pushed new commits but it is still draft because I have to add missing fields from the Frontend Token Definition Schema.

However, I have tested the purpose of this PR with these new commits and it works. The method to check it was using the debugger to see if the `DefaultThemeScopedCSSVariablesProvider` code was executed and by visual inspection of the generated HTML (seeing if a `<style>` node for scope `:root` with default variable values was added to the `<head>`).

File: modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java [hunk 2]
```diff
@@ -23,4 +25,7 @@
 
 	public String getJSON(Locale locale);
 
+	public ParsedFrontendTokenDefinition getParsedFrontendTokenDefinition(
```
Comment (jbalsas) [inline modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: Couldn't we just expose the `getCategories`, `getSets`,  and `getTokens` collection methods and avoid exposing this weird `ParsedTokenDefinition` class?

I would go as far as to not create as an API unless we're planning to mutate the definition in java. I guess it could be useful at that point, but it feels weird for now and we should consider other things if we want to go there.

**Review by jbalsas — COMMENTED** [review]
Hey @izaera, this looks good to me overall... we basically just implement now what we said we wouldn't need 😂 

My main concern is with `ParsedFrontendTokenDefinition`... do we really need that class? What would be its purpose? Are you planning to mutate the definition in any capacity? If not, what's the point of the class?

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java (hunk 1 above)
Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java]: Makes sense to me

File: modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java (hunk 2 above)
Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: you mean limiting the API and not exposing the whole frontend-token-definition schema?

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: or simply removing `ParsedFrontendTokenDefinition` and putting its methods in `FrontendToken` directly?

I created a `ParsedFrontendTokenDefinition` because it implies a parsing operation which is more costly than accessing the JSON directly, but thinking of it again, it is maybe better to remove it...

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: I didn't check the implementation, so I guess you're only parsing the first time it is requested and then caching it?

I don't know how expensive the parsing will be... I was thinking that if we payed the cost upfront, when creating the `FrontenTokenDefinition`, then we wouldn't need to have the `ParsedFrontendTokenDefinition` class and simply expose the fields and accessors we think are valuable from the schema.

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: > simply expose the fields and accessors we think are valuable from the schema.

That is my doubt... 

I have a commit (locally) to expose the full schema to avoid future small PRs to add one tiny thing that will increase the API version and cause overhead. However, it only makes sense if the JSON schema (as it is now) is somewhat final, which I think it is, as this has been released and we have already exposed a public method to retrieve it.

So, is that the case? Can we consider the JSON schema stable enough to release the API now?

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: > Can we consider the JSON schema stable enough to release the API now?

I don't think so.

Additions, though would only cause minor increments, so the overhead should be relatively low. Thus, I'd rather add the minimal API surface possible now (categories, sets, tokens, mappings) and then add more as needed.

One option would be for each entity to also have a `.toJSON()` so even if we miss something, it could be parsed by client code until a proper API is added.

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: :thinking: We already give the full JSON, but it's true that a toJSON() method would be handy (at least you would have the subpart of JSON parsed for you). Let's go with that for now.

Regarding the minimal surface possible, this PR implements that, so I'll push tests for it and stop modeling the rest of the schema.

@victorg1991 do you foresee any other method you will need for sure in the future? (you can see what we have now in [this commit](https://github.com/liferay-frontend/liferay-portal/pull/412/commits/b8bc9a0b42120def2cec7ac0d3cff854310c1223))

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+				layoutSet.getThemeId());
+
+		ParsedFrontendTokenDefinition parsedFrontendTokenDefinition =
+			frontendTokenDefinition.getParsedFrontendTokenDefinition(
+				themeDisplay.getLocale());
+
+		Collection<FrontendTokenCategory> frontendTokenCategories =
+			parsedFrontendTokenDefinition.getFrontendTokenCategories();
+
+		for (FrontendTokenCategory frontendTokenCategory :
```
Comment (jbalsas) [inline modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:75]: Yeah, what I had in mind was, while parsing, to keep the entities separated, so you wouldn't need to loop through categories and sets to get to the tokens.

I think `Collection<FrontendToken> frontendTokenDefinition.getTokens(String locale)` would be very helpful.

Since we've basically parsed the whole thing, that should just mean keeping some additional structures to link entities outside of the root-type level that I guess we have here.

Just to clarify, I'd also keep this, so one can get the sets for a category and the tokens for a set... even the tokens for a category would be useful, but that would be a different API ;)

File: modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java (hunk 2 above)
Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: One more thing, now that I'm modifying the code... We cannot get rid of `ParsedFrontendTokenDefinition` because when parsing we need to define the `Locale` but the raw `FrontendTokenDefinition` is not assigned to any :-(.

However, I propose calling it `LocalizedFrontendTokenDefinition` instead. What do you think @jbalsas ?

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java (hunk 3 above)
Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:75]: So you propose to add helper methods at the root object for direct access, isn't it? I think it's useful, too :+1:

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:75]: yup!

File: modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java (hunk 2 above)
Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: Another option would be propagating the `Locale` parameter to any method that needs it. However that makes the API and the implementation more complex... :thinking:

Comment (victorg1991) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java:28]: I think it could be useful to expose name and label also for the frontend tokens at least, any reason to not do so? Also I think exposing something like getCssVariableName can also be very convenient, in order to avoid looping in the mappings array, what do you think?

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java]: Hey @victorg1991, could you provide a quick recollection of links to code where we could verify the APIs produce better behaviour than right now? As a way to validate? 

We could even apply them here, just to be sure...

**Comment by izaera** [comment]
OK, so I have pushed a new version of the API. Now you can get from almost everywhere to almost everywhere, without needing to traverse the whole graph.

Also I have added `getJSON(Locale)` methods in all hierarchy levels.

Amazingly, everything keeps working correctly.

Now I will add some unit tests to make sure everything is correct.

**Comment by eduardoallegrini** [comment]
Is this fix going to remove the css-custom-properties files?
- https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-theme/frontend-theme-classic/src/css/custom_properties/_custom_properties_set.scss
- https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-theme/frontend-theme-classic/src/css/custom_properties/_custom_properties_variables.scss

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java [hunk 4]
```diff
@@ -65,37 +62,24 @@
...
-
-						String type = frontendTokenMapping.getType();
+			if (!type.equals("cssVariable")) {
+				continue;
+			}
 
-						if (!type.equals("cssVariable")) {
-							continue;
-						}
+			FrontendToken frontendToken =
```
Comment (jbalsas) [inline modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:77]: This inversion of control feels a bit weird... wouldn't it feel more natural to loop through `Token` asking for the mappings?

One convenience method could be `FrontendTokenMapping cssVariableMapping = token.getMapping(FrontendTokenMapping.CSS_VARIABLE)`.

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:77]: This is assuming only one type of mapping, but it could be easily generalized to a Collection.

File: modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java (hunk 2 above)
Comment (victorg1991) [reply modules/apps/frontend-token/frontend-token-definition-api/src/main/java/com/liferay/frontend/token/definition/FrontendTokenDefinition.java]: Sure thing! [Here](https://github.com/liferay/liferay-portal/blob/ef37548663a6635f3be7e8596554c8b1e32c869e/modules/apps/layout/layout-taglib/src/main/java/com/liferay/layout/taglib/internal/display/context/RenderLayoutStructureDisplayContext.java#L863-L881) and [here](https://github.com/liferay/liferay-portal/blob/81fa71c00ea65687f78daffedb903db57f5c47c5/modules/apps/layout/layout-content-page-editor-web/src/main/java/com/liferay/layout/content/page/editor/web/internal/util/StyleBookEntryUtil.java#L40-L62)

File: modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java (hunk 4 above)
Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:77]: I don't see any inversion of control: we have methods to traverse the structure up or down the way you like, and then methods to get all descendants directly (jumping several levels down) so I do it that way to avoid an extra `for`. However, if we add another method to filter by mapping type it can be done with only one `for` as you suggest, of course. 

But, given that we haven't defined mapping types, I'm not sure we want to add that method now. That leads us to another question: are the mapping types defined? If so, where (the JSON schema is missing them)? Seeing the code I thought it was left open for now (and maybe forever, I'm not sure), but maybe that assumption is wrong... :thinking:

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:77]: By "inversion" I meant taking a seemingly isolated "mapping" and then asking for its parent, rather than getting the parents and then querying downwards.

Even if we don't want to have the `filter`, I'd still prefer to get the `tokens` and then go through the mappings. Feels more natural and easier to reason with in the future. Also in case we need something else from the token, we already have the root that we care about in the first step.

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java:77]: Thinking more about the filtering I think it's OK, but we need to decide if the mapping type is a `String` or an `enum`...

If it is a string then nobody defines its possible values (this creates a "problem" when using `cssVariable` because both Echo and us agree on using that...). 

If it is an enum, we (frontend infra) define its valid values, but then it cannot be extended (which can be a limitation or a feature).

Which one seems to be correct? :thinking: 

On the one hand, when themes are used, it looks like mapping types should be normalized (because adding a new type in, say, collaboration, won't work for anything as the base themes won't have that type of mappings). But on the other hand, if we allow using any arbitrary string, people can create their own themes with their own mapping types (for example: for mobile UIs, for any UI system they have, etc...).

I'm not sure what to do because I still haven't 100% clear what other use cases we have in mind...

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java]: I'd say just `String` for now should be okay. Maybe just expose the `CSS_VARIABLE` string as a constant so we can all have a safe way to refer to it?

Comment (izaera) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java]: Given that we are the ones "defining" the CSS variables thing (because it's in **OUR** themes) it makes sense that we define the `CSS_VARIABLE` constant. 

However, I don't think `frontend-token-definitions-api` is a good place for it, because it is far from themes.

But: I can't think of any better place, so I'd say let's put it inside `FrontendTokenMapping` for now.

Maybe the purest way to do it would be adding some OSGi module to hold the `DefaultThemeScopedCSSVariablesProvider` and the constant, that takes care of the frontend tokens stuff related to themes and css variable, but looks a bit overkill (at least for now; and also is quite unclear what will happen in the future, I think).

To sum up: I will put it in `FrontendTokenMapping` unless you can think of a better way to do it.

Comment (jbalsas) [reply modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java]: Sounds good to me

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Hey @eduardoallegrini !

> Is this fix going to remove the css-custom-properties files?

Nope. What made you think it would?

**Review by jbalsas — COMMENTED** [review]
Since this is my own PR, GitHub says I can't approve it 😂 

I checked her on GitHub UI and I think this looks good to go. 👍

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
The failure is not related to the PR, I think, so I'll rerun the tests :roll_eyes:

**Comment by izaera** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
ci:forward

**Comment by jbalsas** [comment]
I think this is safe in case it gets backported, isn't it? We can let the fate of this PR be dictated by the pagan deities of the backportability 😂

**Comment by jbalsas** [comment]
Sigh... @izaera, care to rebase and push-force to see if we can forward this? 

> Target exception: java.lang.RuntimeException: Unable to rebase jbalsas-412-LPS-121610 to master
error: Failed to merge in the changes.

**Comment by izaera** [comment]
ci:forward

**Comment by jbalsas** [comment]
@john-co, @PablitoBonito could you validate these failures? Seem unrelated and maybe present in other PRs...

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated. :heavy_check_mark: 

Looks like the same flaky issue as: https://liferay.slack.com/archives/CLCD3DQLF/p1602610778043500?thread_ts=1602608903.043300&cid=CLCD3DQLF

**Comment by john-co** [comment]
ci:forward

**Comment by izaera** [comment]
ci:forward
